### PR TITLE
chatgpt-retrieval-plugin: init at unstable-2023-03-28

### PR DESCRIPTION
**Pull request by happysalada**
###### Description of changes

This adds quite a bit of pakages unfortunately.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
@MayNiklas I've cheated and disabled the tests for now on tiktoken, to get it available quicker.

File: pkgs/development/python-modules/tiktoken/default.nix [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  };
+  postPatch = ''
+    cp ${./Cargo.lock} Cargo.lock
+  '';
+in
+buildPythonPackage {
+  inherit pname version src postPatch;
+  format = "setuptools";
+
+  disabled = pythonOlder "3.6";
```
Comment (MayNiklas) [inline pkgs/development/python-modules/tiktoken/default.nix:28]: tiktoken requires Python >= 3.8.
See [pyproject.toml](https://github.com/openai/tiktoken/blob/e1c661edf3604706bb2db59cfc7bf92f73c09761/pyproject.toml#L10).

File: pkgs/development/python-modules/tiktoken/default.nix [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  propagatedBuildInputs = [
+    requests
+    regex
+    blobfile
+  ];
+
+  meta = with lib; {
+    description = "tiktoken is a fast BPE tokeniser for use with OpenAI's models.";
+    homepage = "https://github.com/openai/tiktoken";
+    license = licenses.asl20;
```
Comment (MayNiklas) [inline pkgs/development/python-modules/tiktoken/default.nix:56]: I found a different license in the repository:
https://github.com/openai/tiktoken/blob/main/LICENSE

File: pkgs/development/python-modules/tiktoken/default.nix [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+, rustPlatform
+, setuptools-rust
+, libiconv
+, requests
+, regex
+, blobfile
+}:
+let
+  pname = "tiktoken";
+  version = "0.2.0";
```
Comment (MayNiklas) [inline pkgs/development/python-modules/tiktoken/default.nix:15]:
```suggestion
  version = "0.3.3";
```
Is there a reason you chose an older version?

File: pkgs/development/python-modules/tiktoken/default.nix [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-30Gj1HhJl1e1sy6uXpdlfPFZ2NnmdkBJ3Xw6u0nhtA8=";
+  };
```
Comment (MayNiklas) [inline pkgs/development/python-modules/tiktoken/default.nix:19]:
```suggestion
  src = fetchFromGitHub {
    owner = "openai";
    repo = pname;
    rev = "refs/tags/${version}";
    hash = "sha256-5YucJjoYAUTRy7oJ3yb2uE1UZDLIPXnV0bJIDgwGOeA=";
  };
```

We could (!) also use fetchFromGitHub - but that would be more of a style choice from me.

**Review by MayNiklas — CHANGES_REQUESTED** [review]
I only took a look at tiktoken, since I also packaged it today ([my-package](https://github.com/MayNiklas/nixpkgs/blob/8bc84b5aefadee1e7fc8d8d3176fff16b20ecb0a/pkgs/development/python-modules/tiktoken/default.nix)).
I commented what caught my eye (I'm not 100% sure myself on some details).

I personally wasn't able to get the tests to pass under Nix, but everything worked:
-> I also would disable the tests.

File: pkgs/development/python-modules/tiktoken/default.nix (hunk 3 above)
Comment (happysalada) [reply pkgs/development/python-modules/tiktoken/default.nix]: originally I wanted to keep the one that chatgpt-retrieval-plugin was using, but it's working with the most recent version as well so all good.

File: pkgs/development/python-modules/tiktoken/default.nix (hunk 4 above)
Comment (happysalada) [reply pkgs/development/python-modules/tiktoken/default.nix]: I'll pass on that style change. Feel free to change it later if you feel strongly about it.

**Comment by happysalada** [comment]
@MayNiklas I think I've got your main comments covered.

**Comment by happysalada** [comment]
alright the ci is passing, I'm going to go ahead and merge, but happy to have another go if someone sees things worth improving.

**Pull request merged**